analytical_engine/apps/assortativity/average_degree_connectivity/average_degree_connectivity.h

```diff
@@ -0,0 +1,248 @@
+/** Copyright 2020 Alibaba Group Holding Limited.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Author: Ning Xin
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_AVERAGE_DEGREE_CONNECTIVITY_AVERAGE_DEGREE_CONNECTIVITY_H_
+#define ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_AVERAGE_DEGREE_CONNECTIVITY_AVERAGE_DEGREE_CONNECTIVITY_H_
+
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "grape/grape.h"
+
+#include "apps/assortativity/average_degree_connectivity/average_degree_connectivity_context.h"
+#include "core/app/app_base.h"
+#include "core/utils/app_utils.h"
+#include "core/worker/default_worker.h"
+
+namespace gs {
+/**
+ * @brief Compute the average degree connectivity for graph.
+ * @tparam FRAG_T
+ */
+template <typename FRAG_T>
+class AverageDegreeConnectivity
+    : public AppBase<FRAG_T, AverageDegreeConnectivityContext<FRAG_T>>,
+      public grape::Communicator {
+ public:
+  INSTALL_DEFAULT_WORKER(AverageDegreeConnectivity<FRAG_T>,
+                         AverageDegreeConnectivityContext<FRAG_T>, FRAG_T)
+  static constexpr grape::MessageStrategy message_strategy =
+      grape::MessageStrategy::kAlongEdgeToOuterVertex;
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+  using oid_t = typename fragment_t::oid_t;
+  using edata_t = typename fragment_t::edata_t;
+  using pair_msg_t = typename std::pair<int, double>;
+  using degree_connectivity_t =
+      typename std::unordered_map<int, std::pair<double, double>>;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    // process single node
+    if (frag.GetTotalVerticesNum() == 1) {
+      std::vector<size_t> shape{1, 2};
+      std::vector<double> data = {0.0, 0.0};
+      ctx.assign(data, shape);
+      messages.ForceTerminate("single node");
+    }
+    auto inner_vertices = frag.InnerVertices();
+    for (auto& v : inner_vertices) {
+      traverseVertex(v, frag, ctx, messages);
```

Comment:
Seems you process the vertex degree and neighbors degree separately, but actually they can process the same time ? just compute the every vertex related degrees and finally we just need to merge them?

Reply:
and the compute degree phase can be parallel I think